shared/manifest.go

```diff
@@ -68,27 +71,95 @@ func (m *Manifest) unmarshalAll() error {
 	return nil
 }
 
-// Contains checks whether m contains the path.
-func (m *Manifest) Contains(path string) (bool, error) {
-	if m.imap == nil {
-		if err := m.unmarshalAll(); err != nil {
-			return false, err
-		}
+func trieContains(t interface{}, parts []string) interface{} {
+	if len(parts) == 0 {
+		return t
 	}
-	parts := strings.Split(path, "/")
-	// Split always returns at least one element.
-	// Remove the leading empty part.
-	if parts[0] == "" {
-		parts = parts[1:]
+
+	// t could be nil (e.g. if the previous part does not exist in the map), in which case casting will fail.
+	trie, ok := t.(map[string]interface{})
+	if !ok {
+		return nil
+	}
+	return trieContains(trie[parts[0]], parts[1:])
+}
+
+// ContainsFile checks whether m contains a file path (including directories).
+func (m *Manifest) ContainsFile(path string) (bool, error) {
+	if err := m.unmarshalAll(); err != nil {
+		return false, err
 	}
+
+	path = strings.Trim(path, "/")
+	if path == "" {
+		// Root directory always exists.
+		return true, nil
+	}
+	parts := strings.Split(path, "/")
 	for _, items := range m.imap {
-		if trieContains(items, parts) {
+		if trieContains(items, parts) != nil {
 			return true, nil
 		}
 	}
 	return false, nil
 }
 
+// ContainsTest checks whether m contains a full test URL.
+func (m *Manifest) ContainsTest(testURL string) (bool, error) {
+	if err := m.unmarshalAll(); err != nil {
+		return false, err
+	}
+
+	// URLs in the manifest do not include the leading slash (url_base).
```

Comment:
I would suggest extracting this part:

```
	parts, query := ParseTestURL(testURL) 
	for _,  items := range m.imap {
		...
```

Reply:
On second thought, I think this should be folded into `RecoverTestFilename` (and of course it needs to be renamed). A sane API should encapsulate all the details about variants (query strings) and multi-global tests; the caller only needs to in a test URL and gets all the necessary information.

Reply:
So this is fine, but I would personally roll in the trimming of testURL and splitting of path into query. Your call though.